File: scylla-server-go/services/location_service/location_service.go
package location_service

import (
	"log"
	"scylla-server/prisma"
	"scylla-server/prisma/db"
	"scylla-server/services/run_service"
)

// Gets all the locations, or returns nil and the error
func GetAllLocations() ([]db.LocationModel, error) {
	client, ctx := prisma.PrismaClient()

	locations, err := client.Location.FindMany().Exec(ctx)
	if err != nil {
		log.Printf("Err %q\n", err.Error())
		return nil, err
	}

	return locations, nil

}

// Upserts a location and returns it, or returns nil and the error
func UpsertLocation(name string, latitude float64, longitude float64, radius float64, runId int) (*db.LocationModel, error) {
	client, ctx := prisma.PrismaClient()

	location, err := client.Location.UpsertOne(db.Location.Name.Equals(name)).Update(db.Location.Latitude.Set(latitude), db.Location.Longitude.Set(longitude), db.Location.Radius.Set(radius)).Create(db.Location.Name.Set(name), db.Location.Latitude.Set(latitude), db.Location.Longitude.Set(longitude), db.Location.Radius.Set(radius)).Exec(ctx)
	if err != nil {
		log.Printf("Err %q\n", err.Error())
		return nil, err
	}

	// TODO is this needed?
	run_service.GetRunById(runId)

	_, err = client.Run.FindUnique(db.Run.ID.Equals(runId)).Update(db.Run.Location.Link(db.Location.Name.Equals(name))).Exec(ctx)
	if err != nil {
		log.Printf("Err %q\n", err.Error())
		return nil, err
	}

	return location, nil
}

File: scylla-server-rust/src/services/location_service.rs
use prisma_client_rust::QueryError;

use crate::{
    prisma::{self},
    Database,
};

prisma::location::select! {public_location{
    name
    latitude
    longitude
    radius
    runs: select {
        id
        location_name
        driver_name
        system_name
        time
    }
}}

/// Gets all locations
/// * `db` - The prisma client to make the call to
/// returns: A result containing the data or the QueryError propogated by the db
pub async fn get_all_locations(db: &Database) -> Result<Vec<public_location::Data>, QueryError> {
    db.location()
        .find_many(vec![])
        .select(public_location::select())
        .exec()
        .await
}

/// Upserts a location, either creating or updating one depending on its existence
/// * `db` - The prisma client to make the call to
/// * `name` - The name of the location to upsert
/// * `latitude` - The latitude of the location
/// * `longitude` - The longitude of the location
/// * `radius` - The radius of the locations bounds
/// * `run_id` - The run at the location, must already exist!
/// returns: A result containing the data or the QueryError propogated by the db
pub async fn upsert_location(
    db: &Database,
    name: String,
    latitude: f64,
    longitude: f64,
    radius: f64,
    run_id: i32,
) -> Result<public_location::Data, QueryError> {
    let loc = db
        .location()
        .upsert(
            prisma::location::name::equals(name.clone()),
            prisma::location::create(name.clone(), latitude, longitude, radius, vec![]),
            vec![
                prisma::location::latitude::set(latitude),
                prisma::location::longitude::set(longitude),
                prisma::location::radius::set(radius),
            ],
        )
        .select(public_location::select())
        .exec()
        .await?;

    db.run()
        .update(
            prisma::run::id::equals(run_id),
            vec![prisma::run::location::connect(
                prisma::location::name::equals(name),
            )],
        )
        .exec()
        .await?;

    Ok(loc)
}
